### PR TITLE
Intersection selection generated by GRFs instead of the last GRF wins  in GRF exploration phase.

### PR DESCRIPTION
**Pull request by satanson**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

When runtime-bloom filters are evaluated on chunks,  it works in a style of exploring 1 chunk and exploiting next 31 chunks.
In exploration phases, all filters are evaluated and selectivity are computed, only the filters of high selectivity are chosen to be applied to chunks in exploitation phases. selections generated by filters should be intersectioned,  but mistakenly, the last filter(that has largest filter id) overwrite the shared selection, if the last filter has a poor selectivity(a filter in tpcds q85 is just this case, selectivity >0.97), then the chunk is almostly not filtered.

### CODE REVIEW

**Comment by dirtysalt** [comment]
I expect this works for all RuntimeFilter not just GLobal Runtime Filter, right?

File: be/src/exprs/vectorized/runtime_filter_bank.cpp [hunk 1]
```diff
@@ -256,22 +257,31 @@ void RuntimeFilterProbeCollector::do_evaluate(vectorized::Chunk* chunk, RuntimeB
...
         for (auto& kv : eval_context.selectivity) {
             RuntimeFilterProbeDescriptor* rf_desc = kv.second;
             const JoinRuntimeFilter* filter = rf_desc->runtime_filter();
-            if (filter == nullptr) continue;
+            if (filter == nullptr) {
+                continue;
+            }
             ColumnPtr column = rf_desc->probe_expr_ctx()->evaluate(chunk);
-            vectorized::Column::Filter& selection = filter->evaluate(column.get(), &eval_context.running_context);
+            auto true_count = filter->evaluate(column.get(), &eval_context.running_context);
```
Comment (kangkaisen) [inline be/src/exprs/vectorized/runtime_filter_bank.cpp:271]: true_count could be cumulative for multi filters. then we have more chance to early return if true_count ==0, and we needn't has_filtered flag.

File: be/src/exec/exec_node.cpp [hunk 2]
```diff
@@ -163,6 +163,7 @@ Status ExecNode::init_join_runtime_filters(const TPlanNode& tnode, RuntimeState*
             register_runtime_filter_descriptor(state, rf_desc);
         }
     }
+    _runtime_filter_collector.set_plan_node_id(_id);
```
Comment (kangkaisen) [inline be/src/exec/exec_node.cpp:166]: why need to set plan node is here?

**Comment by kangkaisen** [comment]
would better run a daily test.

**Comment by kangkaisen** [comment]
> I expect this works for all RuntimeFilter not just GLobal Runtime Filter, right?

Yes.

File: be/src/exprs/vectorized/runtime_filter.h [hunk 3]
```diff
@@ -403,39 +403,49 @@ class RuntimeBloomFilter final : public JoinRuntimeFilter {
...
-                        _selection[i] = test_data_with_hash(input_data[i], _hash_values[i]);
-
+                        const auto hit = test_data_with_hash(input_data[i], _hash_values[i]);
+                        _selection[i] &= hit;
+                        true_count += hit;
                     } else {
-                        _selection[i] = test_data(input_data[i]);
+                        const auto hit = test_data(input_data[i]);
+                        _selection[i] &= hit;
+                        true_count = hit;
```
Comment (ZiheLiu) [inline be/src/exprs/vectorized/runtime_filter.h:429]: Maybe `true_count += hit`?

File: be/src/exec/exec_node.cpp (hunk 2 above)
Comment (satanson) [reply be/src/exec/exec_node.cpp:166]: we can set condtional breakpoints in gdb when we only focus on runtime filters on a certain plan node.

File: be/src/exprs/vectorized/runtime_filter.h (hunk 3 above)
Comment (satanson) [reply be/src/exprs/vectorized/runtime_filter.h:429]: ok

File: be/src/exprs/vectorized/runtime_filter_bank.cpp (hunk 1 above)
Comment (satanson) [reply be/src/exprs/vectorized/runtime_filter_bank.cpp:271]: ok

File: be/src/exprs/vectorized/runtime_filter_bank.cpp [hunk 4]
```diff
@@ -263,22 +264,29 @@ void RuntimeFilterProbeCollector::do_evaluate(vectorized::Chunk* chunk, RuntimeB
...
             RuntimeFilterProbeDescriptor* rf_desc = kv.second;
             const JoinRuntimeFilter* filter = rf_desc->runtime_filter();
-            if (filter == nullptr) continue;
+            if (filter == nullptr) {
+                continue;
+            }
             ColumnPtr column = rf_desc->probe_expr_ctx()->evaluate(chunk);
-            vectorized::Column::Filter& selection = filter->evaluate(column.get(), &eval_context.running_context);
+            filter->evaluate(column.get(), &eval_context.running_context);
+            true_count = SIMD::count_nonzero(selection);
```
Comment (ZiheLiu) [inline be/src/exprs/vectorized/runtime_filter_bank.cpp:279]: `filter->evaluate` has already returned `true_count`, so maybe we needn't `SIMD::count_nonzero`?
```suggestion
            true_count = filter->evaluate(column.get(), &eval_context.running_context);
```

**Review by ZiheLiu — DISMISSED** [review]

**Comment by stdpain** [comment]
Suppose we have four filters F1 F2 F3 F4 and the filter rate of F1 is 0.5, 0.01, 0.5, 0.5. If we operate with &, then we cannot found a filter rate of 0.01 for F2 and therefore cannot eliminate F2.

**Review by dirtysalt — DISMISSED** [review]

**Review by stdpain — DISMISSED** [review]

**Comment by satanson** [comment]
> Suppose we have four filters F1 F2 F3 F4 and the filter rate of F1 is 0.5, 0.01, 0.5, 0.5. If we operate with &, then we cannot found a filter rate of 0.01 for F2 and therefore cannot eliminate F2.

in exploration phase, all RFs are evaluated, the selectivity is computed inside of do_evaluate respectively, F2 can be found out.

**Review by ZiheLiu — DISMISSED** [review]

**Review by stdpain — DISMISSED** [review]

**Review by kangkaisen — DISMISSED** [review]

**Review by dirtysalt — APPROVED** [review]

**Review by ZiheLiu — APPROVED** [review]

**Review by stdpain — APPROVED** [review]

**Pull request merged**